// appium-geckodriver/build/lib/desired-caps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.desiredCapConstraints = void 0;
const constants_1 = require("./constants");
const desiredCapConstraints = {
    browserName: {
        isString: true
    },
    browserVersion: {
        isString: true
    },
    acceptInsecureCerts: {
        isBoolean: true
    },
    pageLoadStrategy: {
        isString: true
    },
    proxy: {
        isObject: true
    },
    setWindowRect: {
        isBoolean: true
    },
    timeouts: {
        isObject: true
    },
    unhandledPromptBehavior: {
        isString: true
    },
    systemPort: {
        isNumber: true
    },
    marionettePort: {
        isNumber: true
    },
    verbosity: {
        isString: true,
        inclusionCaseInsensitive: Object.values(constants_1.VERBOSITY)
    },
    androidStorage: {
        isString: true,
        inclusionCaseInsensitive: ['auto', 'app', 'internal', 'sdcard']
    },
    'moz:firefoxOptions': {
        isObject: true
    }
};
exports.desiredCapConstraints = desiredCapConstraints;
