// appium-geckodriver/build/lib/driver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const driver_1 = require("appium/driver");
const gecko_1 = __importDefault(require("./gecko"));
const desired_caps_1 = require("./desired-caps");
const index_1 = __importDefault(require("./commands/index"));
const utils_1 = require("./utils");
/** @type {import('@appium/types').RouteMatcher[]} */
const NO_PROXY = [
    ['GET', new RegExp('^/session/[^/]+/appium')],
    ['POST', new RegExp('^/session/[^/]+/appium')],
    ['POST', new RegExp('^/session/[^/]+/element/[^/]+/elements?$')],
    ['POST', new RegExp('^/session/[^/]+/elements?$')],
];
class GeckoDriver extends driver_1.BaseDriver {
    constructor(opts = {}) {
        // @ts-ignore TODO: make args typed
        super(opts);
        this.desiredCapConstraints = desired_caps_1.desiredCapConstraints;
        this.locatorStrategies = [
            'xpath',
            'tag name',
            'link text',
            'partial link text',
            'css selector',
            // Let these two reach Gecko Driver and fail there with a proper error message
            'id',
            'name',
        ];
        this.resetState();
        for (const [cmd, fn] of lodash_1.default.toPairs(index_1.default)) {
            GeckoDriver.prototype[cmd] = fn;
        }
    }
    resetState() {
        this.gecko = null;
        this.proxyReqRes = null;
        this.isProxyActive = false;
    }
    proxyActive() {
        return this.isProxyActive;
    }
    getProxyAvoidList() {
        return NO_PROXY;
    }
    canProxy() {
        return true;
    }
    // @ts-ignore TODO: make args typed
    async createSession(...args) {
        // @ts-ignore TODO: make args typed
        const [sessionId, caps] = await super.createSession(...args);
        this.gecko = new gecko_1.default(this.log, caps);
        try {
            await this.gecko.start((0, utils_1.formatCapsForServer)(caps));
        }
        catch (e) {
            await this.deleteSession();
            throw e;
        }
        this.proxyReqRes = this.gecko.proxy?.proxyReqRes.bind(this.gecko.proxy);
        this.isProxyActive = true;
        return [sessionId, caps];
    }
    async deleteSession() {
        this.log.info('Ending Gecko Driver session');
        await this.gecko?.stop();
        this.resetState();
        await super.deleteSession();
    }
}
exports.default = GeckoDriver;
