// appium-device-farm/lib/src/device-managers/NodeDevices.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const logger_1 = __importDefault(require("../logger"));
class NodeDevices {
    constructor(host) {
        this.host = host;
    }
    postDevicesToHub(devices, arg) {
        return __awaiter(this, void 0, void 0, function* () {
            // DeviceWithPath -> new device
            // DeviceUpdate -> removed device
            logger_1.default.info(`Updating remote android devices ${this.host}/device-farm/api/register`);
            try {
                const status = (yield axios_1.default.post(`${this.host}/device-farm/api/register`, devices, {
                    params: {
                        type: arg,
                    },
                })).status;
                if (status === 200) {
                    if (arg === 'add') {
                        logger_1.default.info(`Pushed devices to hub ${JSON.stringify(devices)}`);
                    }
                    else {
                        logger_1.default.info(`Removed device and pushed information to hub ${JSON.stringify(devices)}`);
                    }
                }
                else {
                    logger_1.default.warn('Something went wrong!!');
                }
            }
            catch (error) {
                logger_1.default.error(`Unable to push devices update to hub. Reason: ${error}`);
            }
        });
    }
    unblockDevice(filter) {
        return __awaiter(this, void 0, void 0, function* () {
            logger_1.default.info(`Unblocking device ${this.host}/device-farm/api/unblock`);
            try {
                const status = (yield axios_1.default.post(`${this.host}/device-farm/api/unblock`, filter, {
                    params: {
                        type: 'unblock',
                    },
                })).status;
                if (status === 200) {
                    logger_1.default.info(`Unblocked device with filter: ${filter}`);
                }
                else {
                    logger_1.default.warn('Something went wrong!!');
                }
            }
            catch (error) {
                logger_1.default.error(`Unable to unblock device. Reason: ${error}`);
            }
        });
    }
}
exports.default = NodeDevices;
